File: modules/student-experience/src/main/java/com/campusos/portal/engine/WorkflowSpec.java
package com.campusos.portal.engine;

import com.campusos.portal.domain.RequestState;
import java.util.List;
import java.util.Map;

/**
 * Everything the UI needs to render a workflow it knows nothing about:
 * the on-path steps for the timeline, the off-path (rejection/return) states,
 * which states are terminal, and human labels. Templates read this, never a type switch.
 */
public record WorkflowSpec(
        String label,
        RequestState initial,
        List<Step> steps,
        Map<RequestState, OffPath> offPath,
        Map<RequestState, String> terminalTone,
        Map<RequestState, String> stateLabels,
        Map<RequestState, List<Transition>> edges) {

    public record Step(RequestState key, String label) {}

    public record OffPath(String label, String tone) {}

    public boolean isTerminal(RequestState s) { return terminalTone.containsKey(s); }

    public String labelFor(RequestState s) {
        return stateLabels.getOrDefault(s, s.name());
    }

    public List<Transition> from(RequestState s) {
        return edges.getOrDefault(s, List.of());
    }
}

File: modules/student-experience/src/engine/WorkflowSpec.js
/**
 * Everything the UI needs to render a workflow it knows nothing about: the on-path steps
 * for the timeline, the off-path (rejection/return) states, which states are terminal, and
 * human labels. Templates read this, never a type switch.
 */
export class WorkflowSpec {
  constructor(label, initial, steps, offPath, terminalTone, stateLabels, edges) {
    this.label = label;
    this.initial = initial;
    this.steps = steps;               // [{key, label}]
    this.offPath = offPath;           // { state: {label, tone} }
    this.terminalTone = terminalTone; // { state: tone }
    this.stateLabels = stateLabels;   // { state: label }
    this.edges = edges;               // { state: Transition[] }
  }

  isTerminal(state) { return Object.prototype.hasOwnProperty.call(this.terminalTone, state); }

  labelFor(state) { return this.stateLabels[state] ?? state; }

  from(state) { return this.edges[state] ?? []; }
}
